File: back-end/src/main/java/com/example/backend/service/RouteService.java
package com.example.backend.service;

import com.example.backend.entities.Route;
import com.example.backend.service.base.BaseService;

import java.util.List;
import java.util.Optional;

public interface RouteService extends BaseService<Route> {
    List<Route> findBySchoolId(Long schoolId);
    List<Route> findByType(Route.RouteType type);
    Optional<Route> findByIdWithStops(Long id);
    List<Route> findBySchoolAndType(Long schoolId, Route.RouteType type);
    
    /**
     * Adds a stop to a route
     * 
     * @param routeId The ID of the route to add the stop to
     * @param stopName The name of the stop
     * @param address The address of the stop (optional)
     * @param latitude The latitude coordinate
     * @param longitude The longitude coordinate
     * @param sequence The sequence number of the stop in the route
     * @param estimatedMinutesFromStart The estimated minutes from the start of the route (optional)
     */
    void addStop(Long routeId, String stopName, String address, Double latitude, Double longitude, 
                 Integer sequence, Integer estimatedMinutesFromStart);
    
    /**
     * Legacy method for adding a stop to a route
     * 
     * @param routeId The ID of the route to add the stop to
     * @param stopName The name of the stop
     * @param latitude The latitude coordinate  
     * @param longitude The longitude coordinate
     * @param sequence The sequence number of the stop in the route
     */
    void addStop(Long routeId, String stopName, Double latitude, Double longitude, Integer sequence);
    
    void updateStop(Long routeId, Long stopId, String stopName, Double latitude, Double longitude, Integer sequence);
    void removeStop(Long routeId, Long stopId);
    void reorderStops(Long routeId, List<Long> stopIds);
    void assignToSchool(Long routeId, Long schoolId);
    void removeFromSchool(Long routeId);
    List<Route> findActiveRoutesWithStats(Long schoolId);
    long countActiveStudentsOnRoute(Long routeId);
    long countCompletedTripsOnRoute(Long routeId);
    double calculateRouteDistance(Long routeId);
    int estimateRouteDuration(Long routeId);
} 

File: front-end/src/services/RouteService.ts
import { Route } from '../core/entities/route.entity';

export interface RouteService {
  getAllRoutes(): Promise<Route[]>;
  getRouteById(id: number): Promise<Route>;
  createRoute(routeData: any): Promise<Route>;
  updateRoute(id: number, routeData: any): Promise<Route>;
  deleteRoute(id: number): Promise<void>;
  findBySchoolId(schoolId: number): Promise<Route[]>;
  findByType(type: string): Promise<Route[]>;
  findByIdWithStops(id: number): Promise<Route>;
  findBySchoolAndType(schoolId: number, type: string): Promise<Route[]>;
  addStop(routeId: number, stopData: any): Promise<void>;
  updateStop(routeId: number, stopId: number, stopData: any): Promise<void>;
  removeStop(routeId: number, stopId: number): Promise<void>;
  reorderStops(routeId: number, stopIds: number[]): Promise<void>;
  assignToSchool(routeId: number, schoolId: number): Promise<void>;
  removeFromSchool(routeId: number): Promise<void>;
  findActiveRoutesWithStats(schoolId: number): Promise<Route[]>;
  getRouteStatistics(routeId: number): Promise<any>;
  countActiveStudentsOnRoute(routeId: number): Promise<number>;
  countCompletedTripsOnRoute(routeId: number): Promise<number>;
  calculateRouteDistance(routeId: number): Promise<number>;
  estimateRouteDuration(routeId: number): Promise<number>;
} 